persistent/picklecache.py

```diff
@@ -229,36 +296,105 @@ def debug_info(self):
 
     # Helpers
-    def _sweep(self, target):
+
+    # Set to true when a deactivation happens in our code. For
+    # compatibility with the C implementation, we can only remove the
+    # node and decrement our non-ghost count if our implementation
+    # actually runs (broken subclasses can forget to call super; ZODB
+    # has tests for this). This gets set to false everytime we examine
+    # a node and checked afterwards. The C implementation has a very
+    # incestuous relatiounship between cPickleCache and cPersistence:
+    # the pickle cache calls _p_deactivate, which is responsible for
+    # both decrementing the non-ghost count and removing its node from
+    # the cache ring. We're trying to keep that to a minimum, but
+    # there's no way around it if we want full compatibility
+    _persistent_deactivate_ran = False
+
+    @_sweeping_ring
+    def _sweep(self, target, target_size_bytes=0):
         # lock
         node = self.ring.next
-        while node is not self.ring and self.non_ghost_count > target:
-            if node.object._p_state not in (STICKY, CHANGED):
-                node.prev.next, node.next.prev = node.next, node.prev
-                node.object = None
-                self.non_ghost_count -= 1
+        ejected = 0
+
+        while (node is not self.ring
+               and ( self.non_ghost_count > target
+                    or (target_size_bytes and self.total_estimated_size > target_size_bytes))):
+            if node.object._p_state == UPTODATE:
+                # The C implementation will only evict things that are specifically
+                # in the up-to-date state
+                self._persistent_deactivate_ran = False
+
+                # sweeping an object out of the cache should also
+                # ghost it---that's what C does. This winds up
+                # calling `update_object_size_estimation`.
+                # Also in C, if this was the last reference to the object,
+                # it removes itself from the `data` dictionary.
+                # If we're under PyPy or Jython, we need to run a GC collection
+                # to make this happen...this is only noticeable though, when
+                # we eject objects. Also, note that we can only take any of these
+                # actions if our _p_deactivate ran, in case of buggy subclasses.
+                # see _persistent_deactivate_ran
+
+                node.object._p_deactivate()
+                if (self._persistent_deactivate_ran
+                    # Test-cases sneak in non-Persistent objects, sigh, so naturally
+                    # they don't cooperate (without this check a bunch of test_picklecache
+                    # breaks)
+                    or not isinstance(node.object, _SWEEPABLE_TYPES)):
+                    ejected += 1
+                    self.__remove_from_ring(node)
             node = node.next
+        if ejected:
+            # TODO: Only do this on PyPy/Jython?
```

Comment:
On non-refcounted implementations like PyPy, the GC is required both to get the bookkeeping counts to work out but also to fulfill the implied promise of `incrgc`/`minimize` of reducing memory usage.

On refcounted implementations, of course, it would just break cycles and clear weakrefs, etc. The question is, is the potentially reduced memory usage (rings are involved in cycles, right? certainly other ZODB objects are)---a good outcome for `incrgc` and friends---worth the potential performance cost? `incrgc` is called by ZODB's connection at the end of every transaction.

I have no answer.